bigquery/storage/managedwriter/routers.go

```diff
@@ -0,0 +1,397 @@
+	}
+	if sr.multiplex && canMultiplex(writer.StreamName()) {
+		return sr.writerAttachMulti(writer)
+	}
+	// Handle non-multiplex writer.
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+	if pair := sr.exclusiveConns[writer.id]; pair != nil {
+		return fmt.Errorf("writer %q already attached", writer.id)
+	}
+	sr.exclusiveConns[writer.id] = newConnection(sr.pool, simplexConnectionMode)
+	return nil
+}
+
+// multiAttach is the multiplex-specific logic for writerAttach.
+// It should only be called from writerAttach.  We use the same
+// orderAndGrow as watchdog, and simply attach the new writer to
+// the most idle connection.
+func (sr *sharedRouter) writerAttachMulti(writer *ManagedStream) error {
+	sr.multiMu.Lock()
+	defer sr.multiMu.Unlock()
+	// order any existing connections
+	sr.orderAndGrowMultiConns()
+	conn := sr.multiConns[0]
+	sr.multiMap[writer.id] = conn
+	return nil
+}
+
+// orderMultiConns orders the connection slice by current load, and will grow
+// the connections if necessary.
+//
+// Should only be called with R/W lock.
+func (sr *sharedRouter) orderAndGrowMultiConns() {
+	sort.SliceStable(sr.multiConns,
+		func(i, j int) bool {
+			return sr.multiConns[i].curLoad() < sr.multiConns[j].curLoad()
+		})
+	if len(sr.multiConns) == 0 {
+		sr.multiConns = []*connection{newConnection(sr.pool, multiplexConnectionMode)}
+	} else if sr.multiConns[0].isLoaded() && len(sr.multiConns) < sr.maxConns {
+		sr.multiConns = append([]*connection{newConnection(sr.pool, multiplexConnectionMode)}, sr.multiConns...)
+	}
+}
+
+var (
+	// Used by rebalanceWriters to avoid rebalancing if the load difference is within the threshold range.
+	connLoadDeltaThreshold = 1.2
+)
+
+// rebalanceWriters looks for opportunities to redistribute traffic load.
+//
+// This is run as part of a heartbeat, when the connections have been ordered
+// by load.
+//
+// Should only be called with the multiplex mutex r/w lock.
+func (sr *sharedRouter) rebalanceWriters() {
+	mostIdleIdx := 0
+	leastIdleIdx := len(sr.multiConns) - 1
+
+	mostIdleConn := sr.multiConns[0]
+	mostIdleLoad := mostIdleConn.curLoad()
+	if mostIdleConn.isLoaded() {
+		// Don't rebalance if all connections are loaded.
+		return
+	}
+	// only look for rebalance opportunies between different connections.
+	for mostIdleIdx != leastIdleIdx {
+		targetConn := sr.multiConns[leastIdleIdx]
+		if targetConn.curLoad() < mostIdleLoad*connLoadDeltaThreshold {
+			// the load delta isn't significant enough between most and least idle connections
+			// to warrant moving traffic.  Done for this heartbeat.
+			return
+		}
+		numWriters := 0
+		candidateID := ""
+		// Walk the writers to find who all shares the multimap, and pick a writer.
+		// TODO: Revisit if we want to maintain an inverted mapping to make this cheaper at heartbeat.
+		for writerID, conn := range sr.multiMap {
+			if conn == targetConn {
+				numWriters++
+				if candidateID == "" {
+					candidateID = writerID
+				}
+			}
+		}
+		if numWriters == 0 {
```

Comment:
in this case when will the connection be closed?

Reply:
This PR doesn't tackle pool shrinking, only growth.  It'll be closed when the client closes.